### PR TITLE
[SYCL] Fix platform selection in opencl-aot

### PR DESCRIPTION
**Pull request by dnmokhov**
Before picking a matched platform name, ensure the device type also
matches. This fixes a mismatch issue that was caused by an identical
supported platform name for CPU and GPU.

Only iterate through available platforms. When the error message is
non-empty, always return CLErr, so that the message is shown.

Signed-off-by: Dmitri Mokhov <dmitri.n.mokhov@intel.com>

### CODE REVIEW

File: opencl-aot/source/utils.cpp [hunk 1]
```diff
@@ -179,7 +179,7 @@ getOpenCLPlatform(DeviceType Type) {
   std::string PlatformName;
 
   const cl_uint MaxPlatformsCount = 10;
```
Comment (romanovvlad) [inline opencl-aot/source/utils.cpp:181]: Now it is not max, right?
Suggest removing the var and create `Platforms` without size passed.

Comment (dnmokhov) [reply opencl-aot/source/utils.cpp:181]: It is still max. On line 186 below,

- `MaxPlatformsCount` must be specified when querying for available platforms;
- `Platforms.data()` array must provide `MaxPlatformsCount` elements;
- `PlatformsCount` cannot be > `MaxPlatformsCount`.

Comment (dnmokhov) [reply opencl-aot/source/utils.cpp]: Removed on second thought.

File: opencl-aot/source/utils.cpp [hunk 2]
```diff
@@ -240,6 +246,7 @@ getOpenCLPlatform(DeviceType Type) {
          DeviceTypesToSupportedPlatformNames[Type]) {
       ErrorMessage += "  " + SupportedPlatformName + '\n';
     }
+    CLErr = OPENCL_AOT_DEVICE_ID_IS_EMPTY;
```
Comment (AlexeySachkov) [inline opencl-aot/source/utils.cpp:249]: Are you sure that this is correct? I feel like this should be set only if we haven't found any platform, i.e. if `!ErrorMessage.empty()` by some reason.

One more minor comment: error code doesn't look related to the problem to me - I suggest to add one more error code for this particular situation

Comment (dnmokhov) [reply opencl-aot/source/utils.cpp:249]: > Are you sure that this is correct? I feel like this should be set only if we haven't found any platform, i.e. if `!ErrorMessage.empty()` by some reason.

This is correct. Lines 243-250:
```cpp
if (!ErrorMessage.empty()) {
  ErrorMessage += "Failed to find any of these OpenCL platforms:\n";
  for (const auto &SupportedPlatformName :
       DeviceTypesToSupportedPlatformNames[Type]) {
    ErrorMessage += "  " + SupportedPlatformName + '\n';
  }
  CLErr = OPENCL_AOT_DEVICE_ID_IS_EMPTY;
}
```

> One more minor comment: error code doesn't look related to the problem to me - I suggest to add one more error code for this particular situation

How does `OPENCL_AOT_PLATFORM_NOT_FOUND` sound?

File: opencl-aot/source/utils.cpp [hunk 3]
```diff
@@ -220,7 +225,8 @@ getOpenCLPlatform(DeviceType Type) {
     auto Result =
         std::find(SupportedPlatformNames.begin(), SupportedPlatformNames.end(),
                   PlatformNameOnLoopIteration);
-    if (Result != SupportedPlatformNames.end()) {
+    if (Result != SupportedPlatformNames.end() &&                  // name match
+        !clFailed(std::get<2>(getOpenCLDevice(Platform, Type)))) { // type match
```
Comment (dm-vodopyanov) [inline opencl-aot/source/utils.cpp:229]: Error code from ```getOpenCLDevice``` is ignored, it is better to return it as the last parameter of the tuple.

File: opencl-aot/source/utils.cpp (hunk 2 above)
Comment (dm-vodopyanov) [reply opencl-aot/source/utils.cpp:249]: ```OPENCL_AOT_PLATFORM_NOT_FOUND``` LGTM, you can add it to ```CLErr``` in case if ```ErrorMessage``` is not empty and ```CLErr``` is still ```CL_SUCCESS```

Comment (AlexeySachkov) [reply opencl-aot/source/utils.cpp:249]: > This is correct. Lines 243-250:

Sorry, I missed `{}` after `if (!ErrorMessage.empty())`. Indeed this is correct.

> How does OPENCL_AOT_PLATFORM_NOT_FOUND sound?

Looks good to me as well

File: opencl-aot/source/utils.cpp (hunk 3 above)
Comment (dnmokhov) [reply opencl-aot/source/utils.cpp:229]: `getOpenCLDevice(Platform, Type)` failure means that we cannot retrieve device ID of type `Type` for `Platform` on this iteration.

If this succeeds on a later iteration (platform of this type is found), we will return `CL_SUCCESS`.

If every iteration fails (platform of this type is never found), we will return `OPENCL_AOT_PLATFORM_NOT_FOUND`.

Comment (dm-vodopyanov) [reply opencl-aot/source/utils.cpp:229]: My point is - we shouldn't return generic error code (```OPENCL_AOT_PLATFORM_NOT_FOUND```) in case if something failed on OpenCL side, we should report OpenCL error to the user, like 
```Failed to retrieve OpenCL device ID: -1 (CL_DEVICE_NOT_FOUND)```.

File: opencl-aot/source/utils.cpp [hunk 4]
```diff
@@ -240,6 +246,7 @@ getOpenCLPlatform(DeviceType Type) {
          DeviceTypesToSupportedPlatformNames[Type]) {
       ErrorMessage += "  " + SupportedPlatformName + '\n';
```
Comment (dm-vodopyanov) [inline opencl-aot/source/utils.cpp:247]: It may be also unclear from these error messages that we couldn't find a device of requested type (like CPU). E.g., the OpenCL platform is actually exists, but it doesn't have OpenCL devices with type CPU

File: opencl-aot/source/utils.cpp [hunk 5]
```diff
@@ -221,24 +227,36 @@ getOpenCLPlatform(DeviceType Type) {
...
+    SupportedPlatforms += "  " + Platform + '\n';
   }
-  if (!ErrorMessage.empty()) {
-    ErrorMessage += "Failed to find any of these OpenCL platforms:\n";
-    for (const auto &SupportedPlatformName :
-         DeviceTypesToSupportedPlatformNames[Type]) {
-      ErrorMessage += "  " + SupportedPlatformName + '\n';
+  if (clFailed(CLErr)) {
+    std::map<DeviceType, std::string> DeviceTypeToDeviceTypeName{
+        {cpu, "CPU"}, {gpu, "GPU"}, {fpga_fast_emu, "accelerator"}};
```
Comment (dm-vodopyanov) [inline opencl-aot/source/utils.cpp:245]: In case of FPGA Fast Emu, user requested ```--device=fpga_fast_emu```, so it may be unclear for him what ```accelerator``` is. So it should be renamed to something like```FPGA Fast Emu```
